wagtail/wagtailcore/models.py

```diff
@@ -40,8 +39,86 @@
 
 PAGE_TEMPLATE_VAR = 'page'
 
+SITE_CACHE = {}
+REQUEST_SITE_MATCH_CACHE = {}
+
 
 class SiteManager(models.Manager):
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extract 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple.
+        """
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    @staticmethod
+    def clear_site_cache():
+        global SITE_CACHE
+        global REQUEST_SITE_MATCH_CACHE
+        SITE_CACHE = {}
+        REQUEST_SITE_MATCH_CACHE = {}
+
+    def _get_by_id(self, site_id):
+        if site_id not in SITE_CACHE:
+            site = self.get(pk=site_id)
+            SITE_CACHE[site_id] = site
+        return SITE_CACHE[site_id]
+
+    def _cache_and_return_match(self, site, match_cache_key):
+        SITE_CACHE[site.pk] = site
+        REQUEST_SITE_MATCH_CACHE[match_cache_key] = site.pk
+        return SITE_CACHE[site.pk]
+
+    def get_default(self):
+        """Returns the 'default' ``Site`` or raise an exception in no site is
+        set as the default. The result is cached."""
+        key = 'default'
+        if key in REQUEST_SITE_MATCH_CACHE:
+            return self._get_by_id(REQUEST_SITE_MATCH_CACHE['default'])
+
+        match = self.get(is_default_site=True)
+        return self._cache_and_return_match(match, key)
+
+    def get_for_request(self, request):
+        """Return the site responsible for dealing with the supplied
+        ``HttpRequest`` instance. The result is cached."""
+        hostname, port = self.get_hostname_and_port_from_request(request)
+        key = "%s:%s" % (hostname, port)
+
+        if key in REQUEST_SITE_MATCH_CACHE:
+            return self._get_by_id(REQUEST_SITE_MATCH_CACHE[key])
+
+        if hostname:
+            if port:
+                try:
+                    # Try to find a site matching both hostname and port
+                    match = self.get(hostname=hostname, port=port)
+                    return self._cache_and_return_match(match, key)
+                except Site.DoesNotExist:
+                    pass
+
+            try:
+                # If there's only one site matching the hostname, use that,
+                # since there's no ambiguity
+                match = self.get(hostname=hostname)
+                return self._cache_and_return_match(match, key)
+            except(Site.DoesNotExist, Site.MultipleObjectsReturned):
+                pass
```

Comment:
Isn’t it better to return the first site matching `hostname` when there are multiple sites for that hostname, instead of returning the default site? I would put `self.filter(hostname=hostname).first()` and remove `MultipleObjectsReturned`.

Reply:
There are a couple of tests that are directly affected by this:

On the plus side: `wagtail.wagtailcore.tests.test_page_model.TestSiteRouting.test_unrecognised_port_and_default_host_routes_to_default_site` is made slightly more efficient (2 queries instead of 3)

But on the negative:
`wagtail.wagtailcore.tests.test_page_model.TestSiteRouting.test_unrecognised_port_on_known_hostname_routes_to_default_site_if_ambiguity` fails.

The existing pattern was the only one I tried that would pass the tests.

Reply:
OK, if I’m not mistaken, the “fetch all once” solution I suggested would handle these correctly.